### PR TITLE
Add "Generate topic name" button to "Subscribe to topic" dialog

### LINKED ISSUE #453: Add "Generate topic name" button
The web app and documentation should advocate for all topics that aren't "well known names" to have high entropy components. Common cryptographic wisdom is that a 2^-32 event is "not going to happen", so by the birthday bound the appropriate level of entropy is `(32 + log2(num_topics_created)) * 2` bits.

64 bits / 6 bits per base64 character = 10.6 characters of base64 (minimum recommendation: 11)
96 bits / 6 bits per base64 character = 16 characters of base64 - sufficient for servers with 65k random topics
128 bits / 6 bits per base64 character = 21.3 characters of base64 (reasonable maximum recommendation: 21)

 - The web app "Subscribe to topic" dialog should include a button to append a hyphen and 16 base64 characters to what's already typed in the dialog box.
 - https://ntfy.sh/docs/publish/ should say "Pick something with about 96 bits of entropy (16 base64 characters)" instead of "that's not easily guessable" and an interactive component to generate one in the browser.

### PR DESCRIPTION
**Pull request by yardenshoham**
Added a new button. When clicked it'll generate a random alphanumeric string and set it as the topic name.

* Closes #453

# Screenshot
image

# Video
https://user-images.githubusercontent.com/20454870/206441009-2e8098cc-c62d-4e2d-b5e5-95c4578988be.mp4

### CODE REVIEW

File: web/package.json [hunk 1]
```diff
@@ -19,6 +19,7 @@
     "i18next-browser-languagedetector": "^6.1.4",
     "i18next-http-backend": "^1.4.0",
     "js-base64": "^3.7.2",
+    "nanoid": "^4.0.0",
```
Comment (binwiederhier) [inline web/package.json:22]: Adding a new library for generating a random string is insane :-) That can easily be done in JS without a new lib.

File: web/src/components/SubscribeDialog.js [hunk 2]
```diff
@@ -118,7 +131,8 @@ const SubscribePage = (props) => {
                         maxLength: 64,
                         "aria-label": t("subscribe_dialog_subscribe_topic_placeholder")
                     }}
-                />
+                    />
+                <Button onClick={generateTopicName} disabled={props.topic.includes("-")}>{t("subscribe_dialog_subscribe_button_generate_topic_name")}</Button><br />
```
Comment (binwiederhier) [inline web/src/components/SubscribeDialog.js:135]: The button should in the same line as the text field, IMHO.

**Review by binwiederhier — CHANGES_REQUESTED** [review]
Thank you for your contribution! Next time, please ask _before_ writing code if I'd accept it. I don't want people to write code and then have me reject it. This change is small enough so it's fine.

I left some comments. Most prominently, adding libraries is evil, and so for something as trivial as generating a random string we should just do in vanilla JS, or with the libs we already have.

File: web/src/components/SubscribeDialog.js (hunk 2 above)
Comment (yardenshoham) [reply web/src/components/SubscribeDialog.js]: Sure, done

File: web/package.json (hunk 1 above)
Comment (yardenshoham) [reply web/package.json]: Dropped the dependency

File: web/src/components/SubscribeDialog.js [hunk 3]
```diff
@@ -77,6 +89,17 @@ const SubscribePage = (props) => {
...
         props.onSuccess();
     };
+    const generateTopicName = () => {
+        const entropy = randomAlphanumericString();
+        let newTopic = props.topic;
+        if (newTopic) {
+            newTopic += "-" + entropy;
+        } else {
+            const sliceIndex = entropy.length / 2;
+            newTopic = entropy.slice(0, sliceIndex) + "-" + entropy.slice(sliceIndex);
```
Comment (binwiederhier) [inline web/src/components/SubscribeDialog.js:99]: That seems quite random. Why the dash?

**Review by binwiederhier — APPROVED** [review]
Looks good

File: web/src/components/SubscribeDialog.js (hunk 3 above)
Comment (binwiederhier) [reply web/src/components/SubscribeDialog.js:99]: Okay I see it now. Didn't see the `contains(-)` before. 

That's ... unexpected and quite the weird UX. IMHO the "append with a dash" logic is also unexpected. A "Generate" button usually just overwrites the existing thing. And then you don't have to ever disable the button. Just make the button generate a random string. Easy as pie.

**Comment by binwiederhier** [comment]
Awesome thanks

**Comment by wunter8** [comment]
Sorry that in late to the party, but aren't topic names case-sensitive? So the alphabet could (should?) include a-z and A-Z? What about `_` and `-`? I know it doesn't have to be able to produce the full range of valid topics, but I figured I'd comment in case it wasn't considered.

Also, I think it might be good to include a sanity check that the generated topic doesn't start with `up`

**Comment by yardenshoham** [comment]
> Sorry that in late to the party, but aren't topic names case-sensitive? So the alphabet could (should?) include a-z and A-Z? 

If they are case-sensitive, I guess we could add `A-Z` to the alphabet.

> What about `_` and `-`? I know it doesn't have to be able to produce the full range of valid topics, but I figured I'd comment in case it wasn't considered.

What are you suggesting here? Simply adding `_` to the alphabet?

> Also, I think it might be good to include a sanity check that the generated topic doesn't start with `up`

Why?

-----

Generally, I think going for the simplest thing is best

**Comment by wunter8** [comment]
> What are you suggesting here? Simply adding `_` to the alphabet?

`_` and `-` are both valid topic characters, so they could be included in the alphabet as well, yes. 

> > Also, I think it might be good to include a sanity check that the generated topic doesn't start with `up`
> 
> Why?

All topics generated by/for UnifiedPush start with `up`. A user generated topic that starts with `up` acts like normal, but many users add anonymous-write permissions to `up*` topics. So if a randomly generated (and presumably secure) topic is generated that starts with `up`, it might be exposed to (malicious) anonymous writes. They would need to get hold of the randomly generated topic, so it's not likely anonymous writes to the topic would happen frequently, but it's a possibility.

This sanity check certainly isn't a necessity. It would just be out of an abundance of caution

**Pull request merged**